dictKeys and dictItems ops on typed dicts return typed lists

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23270 dictKeys and dictItems ops on typed dicts return typed lists**

Differential Revision: [D16448942](https://our.internmc.facebook.com/intern/diff/D16448942/)

smessmer committed:
dictKeys and dictItems ops on typed dicts return typed lists



Differential Revision: [D16448942](https://our.internmc.facebook.com/intern/diff/D16448942/)

eellison commented:
This was fixed separately in https://github.com/pytorch/pytorch/pull/23267

smessmer committed:
Update on "dictKeys and dictItems ops on typed dicts return typed lists"


Differential Revision: [D16448942](https://our.internmc.facebook.com/intern/diff/D16448942/)

smessmer commented:
I rebased past it

smessmer committed:
Update on "dictKeys and dictItems ops on typed dicts return typed lists"


Differential Revision: [D16448942](https://our.internmc.facebook.com/intern/diff/D16448942/)

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -1693,7 +1693,15 @@ int dictLen(Stack& stack) {
 c10::List<Elem> makeListForDictKeysOrValues(
+    const std::pair<c10::optional<TypePtr>, c10::optional<TypePtr>>& types,
     const std::vector<std::pair<IValue, IValue>>& order) {
+  TORCH_INTERNAL_ASSERT(
```
What is the point of the changes to this function ? it already worked

smessmer replied:
Additional error checking that the lists we're passing around have the correct type

eellison replied:
the typesystem should already ensure that it's the correct type. if we're adding runtime enforcements here shouldn't we be adding a runtime enforcement to every op that involves a list?

eellison replied:
It's not a huge deal either way but I still would prefer not to set the precedent that we don't trust the input types of the statically typed langauge. 

@zdevito thoughts ?

smessmer replied:
to reiterate our offline discussion: What this code **should** do is take the type from the source dict and write them to the target list, same as what `makeGenericListForDictKeysOrValues` is doing.

Unfortunately, only `GenericList` has a constructor that takes a runtime type and can do this. `List<T>` doesn't need that because it infers the runtime type from it's `T` type parameter. So instead of passing in the runtime type to the constructor, this is an assertion to check that the inferred type is the one we expect to be and would have passed in if we had created a `GenericList`.

Note that this is a `TORCH_INTERNAL_ASSERT` and not a `TORCH_CHECK` because this is not a type check that reports errors to users. Static typing should make sure that this never fails. But it can fail if there's a PyTorch bug and I'd rather we catch that here than just ignore it.

zdevito replied:
This is just an assert, so it is ok. It improves the error message that would otherwise happen in the `to` call.

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -1704,8 +1712,12 @@ c10::List<Elem> makeListForDictKeysOrValues(
+    const std::pair<c10::optional<TypePtr>, c10::optional<TypePtr>>& types,
     const std::vector<std::pair<IValue, IValue>>& order) {
-  auto values = c10::impl::GenericList(c10::impl::deprecatedUntypedList());
+  auto type = std::get<Index>(types);
```
same here

smessmer replied:
Here, it's more than just error checking. Without this, the runtime key/value type of the Dict wouldn't propagate into the resulting list.

eellison replied:
We create the type of the list based on the static type. E.g. an optional list of ints with a value of [1] should be a generic list. If it this created an intlist that would be a bug. 

how exactly is the runtime type set ? and as with ^^^ the type system should already enforce the correct runtime value

eellison reviewed (commented):
What is the goal of the PR?

smessmer commented:
Each Dict and List stores a type pointer for their elements at runtime and this PR makes sure that this type information is correctly propagated when calling these ops.

smessmer committed:
Update on "dictKeys and dictItems ops on typed dicts return typed lists"


Differential Revision: [D16448942](https://our.internmc.facebook.com/intern/diff/D16448942/)
